ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,807 @@
+        throws IOException {
+      isOpen = false;
+    }
+
+    /**
+     * @return whether readInto() has been called yet.
+     */
+    boolean isReadCalled() {
+      return readCalled;
+    }
+
+    /**
+     * @return the number of chunks that have been written out to the {@link AsyncWritableChannel}
+     */
+    int getNumChunksWrittenOut() {
+      return numChunksWrittenOut.get();
+    }
+
+    /**
+     * Attempt to write the data associated with the blob to the channel passed in by the caller (if the caller has
+     * done so).
+     */
+    private void maybeWriteToChannel() {
+      // if there are chunks available to be written out, do now.
+      if (firstChunk.isComplete() && readCalled) {
+        while (operationException.get() == null && chunkIndexToBuffer.containsKey(indexOfNextChunkToWriteOut)) {
+          ByteBuffer chunkBuf = chunkIndexToBuffer.remove(indexOfNextChunkToWriteOut);
+          asyncWritableChannel.write(chunkBuf, chunkAsyncWriteCallback);
+          indexOfNextChunkToWriteOut++;
+        }
+        if (operationException.get() != null || numChunksWrittenOut.get() == numChunksTotal) {
+          completeRead();
+        }
+      }
+    }
+
+    /**
+     * Complete the read from this {@link ReadableStreamChannel} by invoking the callback and marking the future.
+     */
+    void completeRead() {
+      if (readIntoCallbackCalled.compareAndSet(false, true)) {
+        streamSize = bytesWritten;
+        readIntoFuture.done(bytesWritten, operationException.get());
+        if (readIntoCallback != null) {
+          readIntoCallback.onCompletion(bytesWritten, operationException.get());
+        }
+      }
+      operationCompleted = true;
+    }
+  }
+
+  /**
+   * GetChunk is used to retrieve and hold a data chunk of a composite blob. An object of this class is initialized
+   * to retrieve one data chunk at a time. Once the associated chunk is successfully retrieved, this object can be
+   * reinitialized and used to retrieve a subsequent chunk.
+   */
+  private class GetChunk {
+    // the operation tracker used to track the operation on the current chunk.
+    private OperationTracker chunkOperationTracker;
+    // whether the operation on the current chunk has completed.
+    private boolean chunkCompleted;
+    // the blob id of the current chunk.
+    private BlobId chunkBlobId;
+    // the index of the current chunk in the overall blob.
+    private int chunkIndex;
+    // the most relevant exception encountered for the current chunk.
+    protected RouterException chunkException;
+    // For a GetChunk, responses may be handled multiple times. Regardless of the successTarget,
+    // the actual body of the response is deserialized only once.
+    protected boolean successfullyDeserialized;
+    // map of correlation id to the request metadata for every request issued for this operation.
+    protected final Map<Integer, GetRequestInfo> correlationIdToGetRequestInfo = new TreeMap<>();
```

Comment:
that is a good choice too. Doesn't make much of a difference here as this data structure will never grow much (bounded by the number of replicas)